File: snippets/0414_third_maximum_number.py
from typing import List


# 1. Sorting
def thirdMaxSort(nums: List[int]) -> int:
    nums = list(set(nums))
    nums.sort(reverse=True)

    return nums[2] if len(nums) >= 3 else nums[0]


# 2. Compare
def thirdMaxCompare(nums: List[int]) -> int:
    first, second, third = float("-inf"), float("-inf"), float("-inf")

    for num in nums:
        if num > first:
            first, second, third = num, first, second
        elif first > num > second:
            second, third = num, second
        elif second > num > third:
            third = num

    return third if third != float("-inf") else first


print(thirdMaxSort([3, 2, 1]))  # 1
print(thirdMaxCompare([3, 2, 1]))  # 1

File: snippets/cpp/0414_third_maximum_number.cpp
#include <vector>
#include <iostream>
#include <limits>

using namespace std;

int thirdMax(vector<int> &nums)
{
    long first = numeric_limits<long>::min(), second = numeric_limits<long>::min(), third = numeric_limits<long>::min();

    for (int num : nums)
    {
        if (num == first || num == second || num == third)
        {
            continue;
        }

        if (num > first)
        {
            third = second;
            second = first;
            first = num;
        }
        else if (num > second)
        {
            third = second;
            second = num;
        }
        else if (num > third)
        {
            third = num;
        }
    }

    return third == numeric_limits<long>::min() ? first : third;
}

int main()
{
    vector<int> nums = {3, 2, 1, 5, 6, 4};
    printf("%d\n", thirdMax(nums)); // 4
}
